pkg/conversion/mapi2capi/aws_test.go

```diff
@@ -16,43 +16,140 @@ limitations under the License.
+		{VirtualName: ptr.To("test"), EBS: &mapiv1.EBSBlockDeviceSpec{VolumeSize: ptr.To(int64(10)), DeleteOnTermination: ptr.To(false)}},
+	}))
+	awsMAPIMachineWithVirtualNameAndMissingEBSConfig = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{VirtualName: ptr.To("test")},
+	}))
+	awsMAPIMachineWithMissingEBSConfig = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{DeviceName: ptr.To("/dev/sdb")},
+	}))
+	awsMAPIMachineWithRootVolumeNotDeleted = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{EBS: &mapiv1.EBSBlockDeviceSpec{VolumeSize: ptr.To(int64(10)), DeleteOnTermination: ptr.To(false)}},
+	}))
+	awsMAPIMachineWithNoDeviceAndVirtualName = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{VirtualName: ptr.To("test"), NoDevice: ptr.To("test"), EBS: &mapiv1.EBSBlockDeviceSpec{VolumeSize: ptr.To(int64(10))}},
+	}))
+
+	// MachineSets configurations.
+	awsMAPIMachineSetBase = machinebuilder.MachineSet().WithProviderSpecBuilder(awsBaseProviderSpec)
+
+	// OCP Infrastructure configurations.
+	infra = &configv1.Infrastructure{
+		Spec:   configv1.InfrastructureSpec{},
+		Status: configv1.InfrastructureStatus{InfrastructureName: "sample-cluster-name"},
+	}
+	infraWithRegion = configbuilder.Infrastructure().AsAWS("sample-cluster-name", awsDefaultRegion)
 )
 
-var _ = Describe("mapi2capi AWS", func() {
-	awsMAPIMachine := machinebuilder.Machine().WithProviderSpecBuilder(machinebuilder.AWSProviderSpec()).WithProviderID("aws:///us-west-2a/i-05442bc41c3df969d").Build()
-	awsMAPIMachineSet := machinebuilder.MachineSet().WithProviderSpecBuilder(machinebuilder.AWSProviderSpec()).Build()
-	awsMAPIMachineSet.Spec.Template.Spec.ProviderID = ptr.To("aws:///us-west-2a/i-05442bc41c3df969d") // TODO: do this in machinebuilder.
-	infra := configv1.Infrastructure{
-		Spec: configv1.InfrastructureSpec{},
-		Status: configv1.InfrastructureStatus{
-			InfrastructureName: "sample-cluster-name",
-		},
+var _ = DescribeTable("mapi2capi AWS convert MAPI Machine",
+	func(m *mapiv1.Machine, i *configv1.Infrastructure, errorsMatcher types.GomegaMatcher, warningsMatcher types.GomegaMatcher) {
+		_, _, warns, err := FromAWSMachineAndInfra(m, i).ToMachineAndInfrastructureMachine()
+		Expect(err).To(errorsMatcher, "should have been able to convert providerSpec to MachineTemplateSpec")
+		Expect(warns).To(warningsMatcher, "should have not warned while converting providerSpec to MachineTemplateSpec")
+	},
+	// Base Case.
+	Entry("With a Base configuration", awsMAPIMachineBase.Build(), infra, BeNil(), BeEmpty()),
+	// Only Error.
+	Entry("With LoadBalancers", awsMAPIMachineWithLBs.Build(), infra, MatchError(ContainSubstring(errLoadbalancersNotSupported)), BeEmpty()),
```

Comment:
I'm not a fan of the constants being used in this test case to match errors like this.

I'd prefer to see the raw string here as I can then see what an end user would see.

I also care about how the errors are constructed beyond the limit of the special text we add in. I want to be able to see that the `field.Invalid` constructed a proper message, something like
```
spec.foo.bar: Invalid Value: 123: value must be less than 100
```

Where currently all you would be checking is the final part of that, `value must be less than 100`

Reply:
Yeah I can see this can be desirable, I'll try to see if there is an easy way to do this while also retaining the err variables bit: https://github.com/openshift/cluster-capi-operator/pull/215#discussion_r1796823009

Otherwise we can switch to a full typed out match.